discovery.go

```diff
@@ -0,0 +1,368 @@
+package pubsub
+
+import (
+	"context"
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"sync"
+	"time"
+)
+
+// discover represents the discovery pipeline.
+// The discovery pipeline handles advertising and discovery of peers
+type discover struct {
+	p *PubSub
+
+	// discovery assists in discovering and advertising peers for a topic
+	discovery discovery.Discovery
+
+	// advertising tracks which topics are being advertised
+	advertising map[string]context.CancelFunc
+
+	// bootstrapped tracks which topics have been bootstrapped
+	bootstrapped map[string]chan struct{}
+
+	bootstrappedMux sync.RWMutex
+
+	// discoverQ handles continuing peer discovery
+	discoverQ chan *discoverReq
+
+	// ongoing tracks ongoing discovery requests
+	ongoing map[string]struct{}
+
+	// done handles completion of a discovery request
+	done chan string
+
+	// peerNotifier processes all peer notifications
+	peerNotifier chan peer.ID
+
+	// tracks the channels subscribed to new peer notifications
+	newPeerNotif map[chan peer.ID]*peerEventTracker
+
+	// setup new peer notifier
+	addPeerNotif chan *reqNewPeerNotifier
+
+	// remove peer notifier
+	rmPeerNotif chan chan peer.ID
+}
+
+// newValidation creates a new validation pipeline
+func newDiscover() *discover {
+	return &discover{
+		advertising:  make(map[string]context.CancelFunc),
+		bootstrapped: make(map[string]chan struct{}),
+		discoverQ:    make(chan *discoverReq, 32),
+		ongoing:      make(map[string]struct{}),
+		done:         make(chan string),
+		peerNotifier: make(chan peer.ID, 32),
+		newPeerNotif: make(map[chan peer.ID]*peerEventTracker),
+		addPeerNotif: make(chan *reqNewPeerNotifier),
+		rmPeerNotif:  make(chan chan peer.ID),
+	}
+}
+
+// Start attaches the discovery pipeline to a pubsub instance and starts event loop
+func (d *discover) Start(p *PubSub) {
+	if d.discovery == nil {
+		return
+	}
+
+	d.p = p
+	go d.discoverLoop()
+}
+
+func (d *discover) discoverLoop() {
+	for {
+		select {
+		case discover := <-d.discoverQ:
+			topic := discover.topic
+
+			d.bootstrappedMux.RLock()
+			bootstrapDone := d.bootstrapped[topic]
+			d.bootstrappedMux.RUnlock()
+			select {
+			case _ = <-bootstrapDone:
```

Comment:
you don't need the dummy variable assignment, just `case <-bootstrapDone:`

Reply:
note that you also need to check whether you got a value, the nil value will trigger as done here.

Reply:
@vyzo I don't think that's correct. nil channels block on read, so we'll go to the default and continue. (https://play.golang.org/p/GTDxZ6uLRzG)